rust/agents/relayer/src/msg/gelato_submitter.rs

```diff
@@ -1,98 +1,231 @@
-use abacus_base::{chains::GelatoConf, InboxContracts};
-use abacus_core::db::AbacusDB;
-use tokio::task::JoinHandle;
-
-use eyre::Result;
+use abacus_base::CoreMetrics;
+use abacus_core::{db::AbacusDB, Encode, Signers};
+use abacus_ethereum::validator_manager::INBOXVALIDATORMANAGER_ABI as ivm_abi;
+use ethers::abi::Token;
+use ethers::types::{Address, U256};
+use ethers_contract::BaseContract;
+use eyre::{bail, Result};
+use gelato::chains::Chain;
+use gelato::fwd_req_call::{ForwardRequestArgs, PaymentType, NATIVE_FEE_TOKEN_ADDRESS};
+use prometheus::{Histogram, IntCounter, IntGauge};
 use tokio::sync::mpsc;
+use tokio::time::{sleep, Duration};
+use tokio::{sync::mpsc::error::TryRecvError, task::JoinHandle};
 use tracing::{info_span, instrument::Instrumented, Instrument};
 
 use super::SubmitMessageArgs;
 
-// TODO(webbhorn): Metrics data.
-// TODO(webbhorn): Pull in ForwardRequestOp logic from prior branch.
+const DEFAULT_MAX_FEE: u32 = 1_000_000_000;
 
-#[allow(dead_code)]
 #[derive(Debug)]
 pub(crate) struct GelatoSubmitter {
     /// Source of messages to submit.
-    rx: mpsc::UnboundedReceiver<SubmitMessageArgs>,
-
-    /// Interface to Inbox / InboxValidatorManager on the destination chain.
-    /// Will be useful in retry logic to determine whether or not to re-submit
-    /// forward request to Gelato, if e.g. we have confirmation via inbox syncer
-    /// that the message has already been submitted by some other relayer.
-    inbox_contracts: InboxContracts,
-
+    pub(crate) messages: mpsc::UnboundedReceiver<SubmitMessageArgs>,
```

Comment:
Fun fact, because the struct is `pub(crate)` you can just use `pub` for the fields. Making it more permissive still won't allow anything beyond the struct visibility itself.

Reply:
Good to know, thanks!